File: USACO/Training/Chapter3/Section3/camelot.cpp
/*
TASK: camelot
LANG: C++
*/

#include <bits/stdc++.h>

#define INF 100000

using namespace std;

int n,m;

vector<int> dx = {2,1,-1,-2,-2,-1,1,2};
vector<int> dy = {1,2,2,1,-1,-2,-2,-1};

int dist[31][31][31][31];

bool in(int x, int y){
   return x >= 1 && x <= n && y >= 1 && y <= m;
}

int chtoi(char ch){
   return ch-'A'+1;
}


int main(){
   ios::sync_with_stdio(false);
   cin.tie(0);
   
   ifstream fin ("camelot.in");
   ofstream fout ("camelot.out");
   
   string s;
   getline(fin,s);
   
   stringstream ss(s);
   ss >> n >> m;
   
   int kingx = -1;
   int kingy = -1;
   
   vector<pair<int,int>> knights;
   while(getline(fin,s)){
      ss = stringstream(s);
      
      string a;
      string b;
      while(getline(ss,a,' ')){
         getline(ss,b,' ');
         int c = chtoi(a[0]);
         int r = stoi(b);
         
         if(kingx == -1){
            kingx = r;
            kingy = c;
         } 
         else {
            knights.emplace_back(r,c);
         }
      }
      
   }
   
   //precompute distances
   for(int k = 1; k <= n; k++){
      for(int j = 1; j <= m; j++){
         for(int h = 1; h <= n; h++){
            for(int g = 1; g <= m; g++){
               dist[k][j][h][g] = INF;
            }
         }
      }
   }
      
   for(int sx = 1; sx <= n; sx++){
      for(int sy = 1; sy <= m; sy++){
         dist[sx][sy][sx][sy] = 0;
         queue<pair<int,int>> q;
         q.emplace(sx,sy);
            
         while(!q.empty()){
            auto p = q.front();
            q.pop();
               
            for(int d = 0; d < 8; d++){
               int newx = p.first+dx[d];
               int newy = p.second+dy[d];
               if(!in(newx,newy)) 
                  continue;
               if(dist[sx][sy][newx][newy] != INF) 
                  continue;
                  
               dist[sx][sy][newx][newy] = dist[sx][sy][p.first][p.second]+1;
               q.emplace(newx,newy);
            }
         }   
      }
   }
      
   
   int answer = INF;
   //pick ending square
   for(int ex = 1; ex <= n; ex++){
      for(int ey = 1; ey <= m; ey++){
         //initial answer
         int sum = 0;
         for(const auto& knight : knights){
            sum += dist[ex][ey][knight.first][knight.second];
         }
         if(sum >= answer) 
            continue;
         answer = min(answer, sum+max(abs(ex-kingx),abs(ey-kingy)));
            
         //pick which knight to meet king
         for(const auto& knight : knights){
            //pick where to meet knight
            for(int mx = 1; mx <= n; mx++){
               for(int my = 1; my <= m; my++){
                  int curanswer = sum-dist[ex][ey][knight.first][knight.second];
                  //king move to meeting square
                  curanswer += max(abs(kingx-mx),abs(kingy-my));
                   //knight move to meeting square, then move to ending square
                  curanswer += dist[knight.first][knight.second][mx][my] + dist[mx][my][ex][ey];
                  answer = min(answer,curanswer);
               }
            }
         }
      }
   }
   
   fout << answer << endl;
   
   
   return 0;
}

File: USACO/Training/Chapter3/Section3/camelot.java
/*
TASK: camelot
LANG: JAVA
*/

import java.io.*;
import java.util.*;
//still TLE
class camelot{
   
   public static int INF = 1000000;

   public static int n,m;
   
   public static int[] dx = new int[]{2,1,-1,-2,-2,-1,1,2};
   public static int[] dy = new int[]{1,2,2,1,-1,-2,-2,-1};
   
   public static void main(String[] args) throws IOException{
      BufferedReader f = new BufferedReader(new FileReader("camelot.in"));
      PrintWriter out = new PrintWriter(new BufferedWriter(new FileWriter("camelot.out")));
      
      StringTokenizer st = new StringTokenizer(f.readLine());
      
      n = Integer.parseInt(st.nextToken());
      m = Integer.parseInt(st.nextToken());
      
      int kingx = -1;
      int kingy = -1;
      
      ArrayList<Point> knights = new ArrayList<Point>();
      
      while(true){
         String s = f.readLine();
         if(s == null) 
            break;
         
         st = new StringTokenizer(s);
         
         while(st.hasMoreTokens()){
            int c = ctoi(st.nextToken().charAt(0));
            int r = Integer.parseInt(st.nextToken());
            
            if(kingx == -1){
               kingx = r;
               kingy = c;
            } else {
               knights.add(new Point(r,c));
            }
         }
      }
      
      //dist[sx][sy][ex][ey][x] fastest way for a knight to go from (sx,sy) to (ex,ey)
      //x = 0 -> doesn't pick up king, x = 1 -> picks up king
      int[][][][][] dist = new int[n+1][m+1][n+1][m+1][2];
      for(int k = 1; k <= n; k++){
         for(int j = 1; j <= m; j++){
            for(int h = 1; h <= n; h++){
               for(int g = 1; g <= m; g++){
                  dist[k][j][h][g][0] = INF;
                  dist[k][j][h][g][1] = INF;
               }
            }
         }
      }
      
      for(int sx = 1; sx <= n; sx++){
         for(int sy = 1; sy <= m; sy++){
            PriorityQueue<Point> pq = new PriorityQueue<Point>();
            dist[sx][sy][sx][sy][0] = 0;
            pq.add(new Point(sx,sy,0,0));
            if(kingx == sx && kingy == sy){
               dist[sx][sy][sx][sy][1] = 0;
               pq.add(new Point(sx,sy,1,0));
            }
            
            while(!pq.isEmpty()){
               Point p = pq.poll();
               
               if(dist[sx][sy][p.x][p.y][p.z] != p.w) continue;
               
               for(int d = 0; d < 8; d++){
                  int newx = p.x+dx[d];
                  int newy = p.y+dy[d];
                  if(!in(newx,newy)) 
                     continue;
                  
                  int newd = dist[sx][sy][p.x][p.y][p.z]+1;
                  if(dist[sx][sy][newx][newy][p.z] > newd) {
                     dist[sx][sy][newx][newy][p.z] = newd;
                     pq.add(new Point(newx,newy,p.z,newd));
                  }
                  
                  if(p.z == 0){
                     newd = dist[sx][sy][p.x][p.y][p.z]+1+Math.max(Math.abs(kingx-newx),Math.abs(kingy-newy));
                     if(dist[sx][sy][newx][newy][1] > newd) {
                        dist[sx][sy][newx][newy][1] = newd;
                        pq.add(new Point(newx,newy,1,newd));
                     }
                  }   
               }
            }
            
         }
      }
      
      int answer = INF;
      
      for(int sx = 1; sx <= n; sx++){
         for(int sy = 1; sy <= m; sy++){
            int sum = 0;
            int minadd = INF;
            for(Point knight : knights){
               sum += dist[sx][sy][knight.x][knight.y][0];  
               minadd = Math.min(minadd,dist[sx][sy][knight.x][knight.y][1]-dist[sx][sy][knight.x][knight.y][0]);
            }
            answer = Math.min(answer,sum + Math.max(Math.abs(kingx-sx),Math.abs(kingy-sy)));
            answer = Math.min(answer,sum + minadd);
         }
      }
      
      out.println(answer);
        
      out.close();
   }
   
   public static boolean in(int x, int y){
      return x >= 1 && x <= n && y >= 1 && y <= m;
   }
   
   public static int ctoi(char ch){
      return ch-'A'+1;
   }
   
   public static class Point implements Comparable<Point>{
      int x;
      int y;
      int z;
      int w;
      public Point(int a, int b){
         x = a;
         y = b;  
      }
      public Point(int a, int b, int c,int d){
         x = a;
         y = b;
         z = c;
         w = d;
      }
      
      public int compareTo(Point p){
         return w-p.w;
      }
      
   }
   
}